Replace Tuples with ValueTuples

stephentoub opened the pull request:
Some help to remove allocation, some just help to avoid pulling in Tuple and likely increasing build size.

(Remaining uses of Tuple in netcoreapp don't appear to be worthwhile changing, e.g. a reference type is needed for atomic replacement.)

cc: @jkotas, @marek-safar

jkotas commented on src/libraries/Common/src/System/Security/Cryptography/Pkcs12Kdf.cs:
```diff
@@ -16,14 +16,14 @@ internal static class Pkcs12Kdf
         // https://tools.ietf.org/html/rfc7292#appendix-B.2
-        private static readonly Dictionary<HashAlgorithmName, Tuple<int, int>> s_uvLookup =
-            new Dictionary<HashAlgorithmName, Tuple<int, int>>
+        private static readonly Dictionary<HashAlgorithmName, KeyValuePair<int, int>> s_uvLookup =
```
This does not really need to be Dictionary given the number of items and where it is used. Array + linear search would work just fine.

stephentoub replied:
I can change it.

jkotas commented on src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/ChainPal.cs:
```diff
@@ -303,25 +301,25 @@ private SafeCreateHandle GetCertsArray(IList<SafeHandle> safeHandles)
                 List<X509Certificate2> certsToRead = new List<X509Certificate2>();
 
-                foreach (var element in elements)
+                for (int i = 0; i < elements.Length; i++)
```
What was wrong with foreach?

stephentoub replied:
Just an unnecessary copy of the other item when just the one item was being used.

xtqqczze replied:
https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBLANgHwAEAGAAkIEYUBuAWACgTyKA6AZRjAFcpsMBPFgGEo/AA4YIAcygBDMQAtBADQCsxAJxCYUDNgBm2MLIwwAznXoNCAZnIAmUkNIBvBqQ+kxvAG4mY5CikAGLQABTunlFhapraugZG/vZopNgAdhgAlADaALqkMLgwALYwmWZZkR5u9FFRADLYZhgAPLFaOnqGxqb2AHykYF1mACoQAEowsgAmpAC8pOkwAO6kTS3t6p0JPcn9YVmW1fX60KRhGRhpC6TE1DethcVlFSwN5VIYCg/YANR/Kp1eo1E4gjzDXRjSbTGYsACCMxmYSKpXKGDMOWweRYAElTCUKEcwZ4AL4ncnAjwnbzYPymQIhaDTMAKCJU+oxbbxbpJPqpK65Aqo14YoFRWrg9bNNodHmJXowAZDEbjKazW7LNYbWXcroK/aHY4cqJnWCyVkXPxQZ5ozJpdK20WVEmgk3gyEYtWwhFIlEvdF4glEyxSylRSmkoA==

marek-safar commented on src/libraries/System.Data.Common/src/System/Data/Common/DataStorage.cs:
```diff
@@ -123,8 +123,8 @@ internal abstract class DataStorage
 
-        private static readonly Func<Type, Tuple<bool, bool, bool, bool>> s_inspectTypeForInterfaces = InspectTypeForInterfaces;
-        private static readonly ConcurrentDictionary<Type, Tuple<bool, bool, bool, bool>> s_typeImplementsInterface = new ConcurrentDictionary<Type, Tuple<bool, bool, bool, bool>>();
+        private static readonly Func<Type, (bool, bool, bool, bool)> s_inspectTypeForInterfaces = InspectTypeForInterfaces;
```
It could also be bitmasked byte

stephentoub replied:
It could.  But that's a larger change than I'd like to make here.  And if this code is being used, I don't expect the size impact is going to be a huge concern.  I'm also not confident in our testing of this area.

stephentoub replied:
(Turns out this file is also compiled into libraries that build for netstandard2.0, so I've reverted the changes here entirely.)

marek-safar commented on src/libraries/System.Net.NameResolution/src/System/Net/Dns.cs:
```diff
@@ -346,7 +346,7 @@ public static IPHostEntry EndResolve(IAsyncResult asyncResult)
                 {
                     IPAddress a => a,
-                    Tuple<IPAddress, AddressFamily> t => t.Item1,
+                    ValueTuple<IPAddress, AddressFamily> t => t.Item1,
```
KeyValuePair instead?

marek-safar commented on src/libraries/System.Net.NameResolution/src/System/Net/Dns.cs:
```diff
@@ -546,17 +546,17 @@ private static Task GetHostEntryOrAddressesCoreAsync(string hostName, bool justR
                 {
                     string h => GetHostAddressesCore(h, AddressFamily.Unspecified),
-                    Tuple<string, AddressFamily> t => GetHostAddressesCore(t.Item1, t.Item2),
+                    ValueTuple<string, AddressFamily> t => GetHostAddressesCore(t.Item1, t.Item2),
```
KeyValuePair instead?

marek-safar commented on src/libraries/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/AesImplementation.Unix.cs:
```diff
@@ -25,32 +25,32 @@ internal partial class AesImplementation
             bool foundKeysize = false;
 
-            foreach (var tuple in s_algorithmInitializers)
+            foreach ((int, int, CipherMode, Func<IntPtr>) tuple in s_algorithmInitializers)
```
Consider replacing the loop with a simple switch to avoid expensive .cctor too

stephentoub replied:
Good suggestion.  Pattern matching for the win :)

stephentoub committed:
Replace Tuples with ValueTuples

Some help to remove allocation, some just help to avoid pulling in Tuple and likely increasing build size.

stephentoub commented on src/libraries/Common/src/System/Security/Cryptography/Pkcs12Kdf.cs:
```diff
@@ -14,17 +14,16 @@ internal static class Pkcs12Kdf
+            (HashAlgorithmName.SHA256, 256, 512),
+            (HashAlgorithmName.SHA384, 384, 1024),
+            (HashAlgorithmName.SHA512, 512, 1024),
+        };
```
@bartonjs, FYI

stephentoub commented on src/libraries/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/AesImplementation.Unix.cs:
```diff
@@ -25,50 +25,29 @@ internal partial class AesImplementation
+                _ => throw (keySize == 128 || keySize == 192 || keySize == 256 ? (Exception)
+                        new NotSupportedException() :
+                        new CryptographicException(SR.Cryptography_InvalidKeySize)),
+            };
```
@bartonjs, FYI

bartonjs commented on src/libraries/Common/src/System/Security/Cryptography/Pkcs12Kdf.cs:
```diff
@@ -85,13 +84,21 @@ internal static class Pkcs12Kdf
+            if (u == -1)
+            {
+                throw new CryptographicException(SR.Cryptography_UnknownHashAlgorithm, hashAlgorithm.Name);
+            }
```
```suggestion
            }

```

stephentoub committed:
Update src/libraries/Common/src/System/Security/Cryptography/Pkcs12Kdf.cs

Co-authored-by: Jeremy Barton <jbarton@microsoft.com>
